packages/notifier/src/publish-kit.js

```diff
@@ -180,12 +200,221 @@ export const makePublishKit = () => {
       advanceCurrent(true, finalValue);
     },
     fail: reason => {
-      const rejection = HandledPromise.reject(reason);
-      // Suppress unhandled rejection error.
-      rejection.catch(() => {});
-      advanceCurrent(true, undefined, rejection);
+      advanceCurrent(true, undefined, makeQuietRejection(reason));
     },
   });
   return harden({ publisher, subscriber });
 };
 harden(makePublishKit);
+
+// TODO: Move durable publish kit to a new file?
+
+/**
+ * @typedef {ReturnType<typeof initDurablePublishKitState>} DurablePublishKitState
+ */
+
+/**
+ * @param {DurablePublishKitOptions} [options]
+ */
+const initDurablePublishKitState = (options = {}) => {
+  const { valueDurability = 'mandatory' } = options;
+  assert.equal(valueDurability, 'mandatory');
+  return {
+    // configuration
+    valueDurability,
+
+    // lifecycle progress
+    publishCount: 0n,
+    status: 'live', // | 'finished' | 'failed'
+
+    // persisted result data
+    hasValue: false,
+    value: undefined,
+  };
+};
+
+/** @type {WeakMap<object, {currentP, tailP, tailR}>} */
+const durablePublishKitEphemeralData = new WeakMap();
+
+// XXX Does state actually work as a WeakMap key?
+/**
+ * Returns the current/next-result promises and next-result resolver
+ * associated with a given state object backing a durable publish kit.
+ * They are lost on upgrade, but recreated on-demand.
+ * Such recreation preserves the value in (but not the identity of) the
+ * current { value, done } result when possible, which is always the
+ * case when that value is terminal (i.e., from `finish` or `fail`) or
+ * when the durable publish kit is configured with
+ * `valueDurability: 'mandatory'`.
+ *
+ * @param {DurablePublishKitState} state
+ */
+const provideDurablePublishKitEphemeralData = state => {
+  const foundData = durablePublishKitEphemeralData.get(state);
+  if (foundData) {
+    return foundData;
+  }
+  const { status, publishCount } = state;
+  let newData;
+  if (status === 'failed') {
+    newData = {
+      currentP: makeQuietRejection(state.value),
```

Comment:
Is it noted somewhere that `state.value` represents a rejection reason if `state.status === 'failed'`? I mention this only because it threw me at first, because the name suggested it would only represent fulfillments, i.e., iterator result values.

Reply:
What do you think of typing `status` as `'failed'`, `'finished'`, or `'live'`?

Reply:
> What do you think of typing `status` as `'failed'`, `'finished'`, or `'live'`?

I'm in favor of that... do you have a preferred location to put the typing? Would something like this work?
```patch
-    status: 'live', // | 'finished' | 'failed'
+    status: /** @type {'live'|'finished'|'failed'} */ ('live'),
```

Reply:
That would work. Though you'd have to do the same for `value`:
<img width="257" alt="Screenshot 2022-12-06 at 10 32 42 AM" src="https://user-images.githubusercontent.com/21505/205993171-c600f3c9-e4f8-4b49-b4fe-9e7c3ed9ea06.png">
That wasn't obviously necessary. So I'd suggest having an explicit type instead, making apparent what values are typed or not. 

It would also have the advantage of being a better place for the documentation on what the values mean. E.g. this comment could be visible when hovering over an instance of the state type: https://github.com/Agoric/agoric-sdk/blob/96979ceb43e7c04e6f93707eb198405859a797d4/packages/notifier/src/publish-kit.js#L229-L232

Reply:
Done, please review: https://github.com/Agoric/agoric-sdk/pull/6502/commits/a0d56611a97894d015b2fc4a9ffa66ea514c0576